modules/stitching/src/exposure_compensate.cpp

```diff
@@ -298,6 +308,73 @@ void GainCompensator::setMatGains(std::vector<Mat>& umv)
     }
 }
 
+void GainCompensator::prepareSimilarityMask(
+    const std::vector<Point> &corners, const std::vector<UMat> &images)
+{
+    if (similarity_threshold_ >= 1)
+    {
+        LOGLN("  skipping similarity mask: disabled");
+        return;
+    }
+    if (!similarities_.empty())
+    {
+        LOGLN("  skipping similarity mask: already set");
+        return;
+    }
+
+    LOGLN("  calculating similarity mask");
+    const int num_images = static_cast<int>(images.size());
+    for (int i = 0; i < num_images; ++i)
+    {
+        for (int j = i; j < num_images; ++j)
+        {
+            Rect roi;
+            if (overlapRoi(corners[i], corners[j], images[i].size(), images[j].size(), roi))
+            {
+                Mat subimg1 = images[i](Rect(roi.tl() - corners[i], roi.br() - corners[i])).getMat(ACCESS_READ);
+                Mat subimg2 = images[j](Rect(roi.tl() - corners[j], roi.br() - corners[j])).getMat(ACCESS_READ);
+                Mat similarity = getSimilarityMask(subimg1, subimg2);
+                similarities_.push_back(similarity);
+            }
+        }
+    }
+}
+
+Mat GainCompensator::getSimilarityMask(Mat src1, Mat src2)
+{
+    CV_Assert(src1.rows == src2.rows && src1.cols == src2.cols);
+    CV_Assert(src1.type() == src2.type());
+    CV_Assert(src1.type() == CV_8UC3 || src1.type() == CV_8UC1);
+
+    Mat similarity(src1.rows, src1.cols, CV_8UC1);
+    for (int y = 0; y < similarity.rows; ++y)
+    {
+        for (int x = 0; x < similarity.cols; ++x)
+        {
+            double diff;
+            if (src1.channels() == 3)
+            {
+                Vec<int, 3> vec_diff =
+                    Vec<int, 3>(*src1.ptr<Vec<uchar, 3>>(y, x))
+                    - Vec<int, 3>(*src2.ptr<Vec<uchar, 3>>(y, x));
+                diff = norm(vec_diff / 255);
```

Comment:
Avoid divisions. Prefer multiplications.

Or move this to `diff <= similarity_threshold_` check (by adjusting threshold).

Reply:
Makes me realize this is wrong, it should be `Vec<float, 3>` the code above will (almost) always round down to 0. I'll fix that as well